### PR TITLE
ref(indexer): allow spanner indexer in prod via indexer_db cli arg

### PR DESCRIPTION
**Pull request by MeredithAnya**
**context**
We want to test out alternative backends for the metrics indexer, and test them in parallel to the current indexer. I've added a cli arg `--indexer-db` that defaults to `postgres`, that can be used to determine which backend we want to use. 

There is no problem with having another consumer group consume from the ingest topic (`ingest-metrics` or `ingest-performance-metrics`). However, we don't want to double produce to the snuba topic so we have to make a new topic to produce to. I've only added the logic for the performance metrics, but the same could be done for release health (aka would need a new dummy topic and logic to override the `output_topic`)

**other changes**
* ~I got rid of the `Db.Key` because it looked like that mapping could be replaced by using the `use_case_id`. Open to feedback on that, can always put it back.~ Got pulled out by @lynnagara  in https://github.com/getsentry/sentry/pull/37790
* I had to update the `WritesLimiter` but I think that makes sense so that we have a separation in redis between the backend implementations 
    * **[edit]** update rate limiter in https://github.com/getsentry/sentry/pull/38192
* I pass the full `MetricsIngestConfiguration` to the `process_messages` now instead of just the `use_case_id`. Seemed to make sense since both of the args I would have been passing in were on the config already
    *  **[edit]:** update the message processor in https://github.com/getsentry/sentry/pull/38225

This _should_ be able to be run in prod provided the ops config for the consumer and topic are added. 

**notes:**
The Last Seen Updater currently only works with the postgres backend as it is coupled to the postgres tables, futures changes would need to be made to enable different backends for the last seen updater https://github.com/getsentry/sentry/blob/fb1bc70261712ce11ee609123d77a4f7065d2ab3/src/sentry/sentry_metrics/consumers/last_seen_updater.py#L128

### CODE REVIEW

File: src/sentry/sentry_metrics/configuration.py [hunk 1]
```diff
@@ -41,23 +36,32 @@ def _register_ingest_config(config: MetricsIngestConfiguration) -> None:
     _METRICS_INGEST_CONFIG_BY_USE_CASE[config.use_case_id] = config
 
 
-def get_ingest_config(use_case_key: UseCaseKey) -> MetricsIngestConfiguration:
+def get_ingest_config(use_case_key: UseCaseKey, db_backend: str) -> MetricsIngestConfiguration:
     if len(_METRICS_INGEST_CONFIG_BY_USE_CASE) == 0:
         _register_ingest_config(
             MetricsIngestConfiguration(
-                db_model=DbKey.STRING_INDEXER,
+                db_backend=db_backend,
```
Comment (fpacifici) [inline src/sentry/sentry_metrics/configuration.py:43]: This is going to cause an issue. 
The `get_ingest_config` is called in multiple places within sentry and the indexer to get a specific indexer config. When you run this method from the indexer code you will always pass the same use case key so no problem. But sentry (being it a monolith) needs to have access to all indexer configs at the same time. So, if you use `db_backend` to initialize the config, you will fundamentally be ignoring that parameter for all subsequent calls. 

If you call `get_ingest_config('release_health', 'spanner')` you get the spanner indexer.
If you then call `get_ingest_config('release_health', 'postgres')`, the db_backend is ignored and the spanner one is returned as the first called has initialized the config. 

Basically you cannot use the parameters passed to the method in the initialization of the config as that is then stored in the module and not changed anymore no matter which parameters you pass.

File: src/sentry/sentry_metrics/configuration.py [hunk 2]
```diff
@@ -41,23 +36,32 @@ def _register_ingest_config(config: MetricsIngestConfiguration) -> None:
...
             )
         )
+
+        # Hack to make sure that if we run backends in parallel we aren't
+        # double producing to the snuba-metrics-generics topic
+        output_topic = (
+            settings.KAFKA_SNUBA_GENERICS_METRICS_DUMMY
+            if db_backend == "cloudspanner"
+            else settings.KAFKA_SNUBA_GENERIC_METRICS
+        )
```
Comment (fpacifici) [inline src/sentry/sentry_metrics/configuration.py:58]: Same as above. Sentry may call `get_ingest_config` subsequently with different DB backends. As soon as that happens both implementations will break.

File: src/sentry/sentry_metrics/configuration.py [hunk 3]
```diff
@@ -19,14 +19,9 @@ def from_str(use_case: str) -> "UseCaseKey":
...
 
-class DbKey(Enum):
-    STRING_INDEXER = "StringIndexer"
-    PERF_STRING_INDEXER = "PerfStringIndexer"
-
-
 @dataclass(frozen=True)
 class MetricsIngestConfiguration:
-    db_model: DbKey
+    db_backend: str
```
Comment (fpacifici) [inline src/sentry/sentry_metrics/configuration.py:24]: I think the `MetricsIngestConfiguration` we designed is not sufficient for this use case and we need to rethink it a bit to do this task in a safe way.

Context on `MetricsIngestConfiguraiton`.
When we introduced the performance indexer (and we managed to break release health) it became clear fairly quickly that there was a number of parameters that had to be specific to each indexer and that we should group them under some structure that would represent an indexer instance to remain sane. Otherwise we would have had a number of independent configurations that we would have had to override. With that solution we were pretty much sure to make mistakes and to cause production issues.
Thus the `metricsIngestConfiguration` was born. The idea was that anywhere you needed an indexer you would reference it through the `use_case_id`, every indexer public method requires a `use_case_id` with no default, so no mistakes can be made and the indexer never reference settings unless through this object. 

The `use_case_id`, thus, is meant to represent the product concept (release health and performance so far). Nothing more.  This is because the product code using the indexer is generally product specific and should not change if we make changes to the indexer storage.

Now this use case (two or more indexers for performance) breaks everything.
I think we may even (by the end of next week) end up running multiple cloud spanner indexers in parallel in prod with different transactions models. 

This PR seems to be overriding specific fields in the `metricsIngestConfiguration` depending on which db_backend we want when we call `get_ingest_config`. 
I think we should instead provide a totally different `MetricsIngestConfiguration` object depending on the db_backend. This would make it easier to customize all its field and provide a more detailed DB configuration. 
See below for more details on which additional fields we should have here.

File: src/sentry/sentry_metrics/consumers/indexer/processing.py [hunk 4]
```diff
@@ -25,8 +25,15 @@ def get_indexer():  # type: ignore
     return indexer
 
 
+@functools.lru_cache(maxsize=10)
+def get_spanner_indexer():  # type: ignore
+    from sentry.sentry_metrics.indexer.cloudspanner import CloudSpannerIndexer
+
+    return CloudSpannerIndexer()
```
Comment (fpacifici) [inline src/sentry/sentry_metrics/consumers/indexer/processing.py:32]: I think may want to run two spanners in parallel. A very real use case could be to run one on a multi region spanner and one on a single region spanner.

File: src/sentry/sentry_metrics/indexer/postgres_v2.py [hunk 5]
```diff
@@ -144,7 +145,9 @@ def bulk_record(
             indexer_cache.set_many(new_results_to_cache, use_case_id.value)
             return cache_key_results.merge(db_read_key_results)
 
-        with writes_limiter.check_write_limits(use_case_id, db_write_keys) as writes_limiter_state:
+        with writes_limiter.check_write_limits(
+            use_case_id, db_write_keys, _INDEXER_DB
+        ) as writes_limiter_state:
```
Comment (fpacifici) [inline src/sentry/sentry_metrics/indexer/postgres_v2.py:150]: I think the rate limiter partition should be an explicit field in `MetricsIngestConfiguration`. (thus two fully different config objects one for spanner performance and one for postgres performance).
The reason is that by using the DB backend you are creating an implicit assumption the correctness of your system depends on: there is one rate limiter per DB backend.
While this is true, it is not intuitive. The rate limiter may be disconnected from the db backend. It owuld be hard to remember about this issue. We messed up the cache key exactly because of a similar issue.
If instead you had two `MetricsIngestConfiguration` and each had to define the rate limiter partition, when writing the config you will be required to think about this problem, thus making it less likely to make mistakes.

File: src/sentry/sentry_metrics/consumers/indexer/processing.py [hunk 6]
```diff
@@ -53,15 +60,18 @@ def process_messages(
     else:
         # This, instead of importing the normal way, was likely done to prevent
         # fork-safety issues.
-        indexer = get_indexer()
+        if config.db_backend == "cloudspanner":
+            indexer = get_spanner_indexer()
+        else:
+            indexer = get_indexer()
```
Comment (fpacifici) [inline src/sentry/sentry_metrics/consumers/indexer/processing.py:66]: How will sentry pick the right indexer to reverse resolve ? 
For this test we do not need to care, but we will have to cut over at a point. Do you envision us doing it through multiple implementations of the service delegator ?

File: src/sentry/sentry_metrics/configuration.py [hunk 7]
```diff
@@ -41,23 +36,32 @@ def _register_ingest_config(config: MetricsIngestConfiguration) -> None:
...
+
         _register_ingest_config(
             MetricsIngestConfiguration(
-                db_model=DbKey.PERF_STRING_INDEXER,
+                db_backend=db_backend,
                 input_topic=settings.KAFKA_INGEST_PERFORMANCE_METRICS,
-                output_topic=settings.KAFKA_SNUBA_GENERIC_METRICS,
+                output_topic=output_topic,
                 use_case_id=UseCaseKey.PERFORMANCE,
                 internal_metrics_tag="perf",
```
Comment (fpacifici) [inline src/sentry/sentry_metrics/configuration.py:66]: You will have to override this field as well otherwise all datadog metrics will be mixed up.

**Review by fpacifici — CHANGES_REQUESTED** [review]
See my comments inline. 

tl;dr I think we should have one `MetricsIngestConfiguration` instance for spanner and one for postgres. Though they would have to share the use case id. So we may still need the db_backend to disambiguate.

File: src/sentry/sentry_metrics/configuration.py (hunk 1 above)
Comment (MeredithAnya) [reply src/sentry/sentry_metrics/configuration.py:43]: ah yeah okay cause `_METRICS_INGEST_CONFIG_BY_USE_CASE[use_case_key]` is using the use_case_key for look up. makes sense.

File: src/sentry/sentry_metrics/consumers/indexer/processing.py (hunk 6 above)
Comment (MeredithAnya) [reply src/sentry/sentry_metrics/consumers/indexer/processing.py]: I think eventually we could load the cloudspanner be loaded via the lazy service wrapper, and ideally `from sentry.sentry_metrics import indexer` can just work. Right now since I'm adding a cli arg, that doesn't seem to mesh well with loading from settings. It also seems to me like if we want to be easily able to cut over or switch between the two, we'd wanna stay with cli arg to be make those changes more quickly. anyway just my initial thoughts, would have to think a bit more deeply on it

File: src/sentry/sentry_metrics/db.py [hunk 8]
```diff
@@ -0,0 +1,2 @@
+POSTGRES_DB = "postgres"
+CLOUDSPANNER_DB = "cloudspanner"
```
Comment (MeredithAnya) [inline src/sentry/sentry_metrics/db.py:2]: if we wanted we could split cloudspanner into like:

```python
CLOUDSPANNER_SINGLE_DB = "cloudspanner-single"
CLOUDSPANNER_MULTI_DB = "cloudspanner-multi"
```

for single and multi cloudspanner instances, assuming that you'd have the `use_case_id` in the configuration be the indicator for the db within the instance

Comment (lynnagara) [reply src/sentry/sentry_metrics/db.py:2]: I would not add this db.py file with these strings. I think you'd be better off with an enum in the configuration.py file like how the rest of the options for running the indexer are defined.

File: src/sentry/sentry_metrics/configuration.py [hunk 9]
```diff
@@ -12,33 +14,39 @@ class UseCaseKey(Enum):
...
     input_topic: str
     output_topic: str
     use_case_id: UseCaseKey
     internal_metrics_tag: Optional[str]
     writes_limiter_cluster_options: Mapping[str, Any]
 
 
-_METRICS_INGEST_CONFIG_BY_USE_CASE: MutableMapping[UseCaseKey, MetricsIngestConfiguration] = dict()
+_METRICS_INGEST_CONFIG_BY_USE_CASE: MutableMapping[
+    Tuple[UseCaseKey, str], MetricsIngestConfiguration
```
Comment (lynnagara) [inline src/sentry/sentry_metrics/configuration.py:26]: How about `DbBackend` enum here instead of a string?

File: src/sentry/sentry_metrics/db.py (hunk 8 above)
Comment (MeredithAnya) [reply src/sentry/sentry_metrics/db.py:2]: okay that works, wasn't too sure about putting these here anyway

File: src/sentry/sentry_metrics/configuration.py [hunk 10]
```diff
@@ -47,4 +55,30 @@ def get_ingest_config(use_case_key: UseCaseKey) -> MetricsIngestConfiguration:
             )
         )
 
-    return _METRICS_INGEST_CONFIG_BY_USE_CASE[use_case_key]
+        # Make sure that if we run backends in parallel we aren't
+        # double producing to the snuba-metrics-generics topic
+        dummy_output_topic = settings.KAFKA_SNUBA_GENERICS_METRICS_DUMMY
```
Comment (lynnagara) [inline src/sentry/sentry_metrics/configuration.py:60]: I'd call this _CLOUDSPANNER instead of _DUMMY. What's the reason for calling it dummy? Are you saying we will not be using it permanently after the migration?

File: src/sentry/sentry_metrics/consumers/indexer/processing.py [hunk 11]
```diff
@@ -53,15 +61,18 @@ def process_messages(
     else:
         # This, instead of importing the normal way, was likely done to prevent
         # fork-safety issues.
-        indexer = get_indexer()
+        if config.db_backend == CLOUDSPANNER_DB:
+            indexer = get_spanner_indexer()
+        else:
+            indexer = get_indexer()
```
Comment (lynnagara) [inline src/sentry/sentry_metrics/consumers/indexer/processing.py:67]:
```suggestion
            indexer = get_postgres_indexer()
```

File: src/sentry/sentry_metrics/configuration.py (hunk 10 above)
Comment (MeredithAnya) [reply src/sentry/sentry_metrics/configuration.py:60]: Well I was thinking we could use the same output topic for testing multiple backends, so if we wanted to try to test something out after cloudspanner it could work for that too.

Yeah in my mind this topic would just be used for testing

File: src/sentry/sentry_metrics/indexer/ratelimiters.py [hunk 12]
```diff
@@ -139,7 +144,9 @@ def check_write_limits(self, use_case_id: UseCaseKey, keys: KeyCollection) -> Ra
         """
 
         org_ids, requests = _construct_quota_requests(use_case_id, keys)
-        timestamp, grants = self._get_rate_limiter(use_case_id).check_within_quotas(requests)
+        timestamp, grants = self._get_rate_limiter(use_case_id, indexer_db).check_within_quotas(
```
Comment (lynnagara) [inline src/sentry/sentry_metrics/indexer/ratelimiters.py:147]: How does the rate limiter namespacing work? Looking at sliding_windows.py the redis key and cluster seems to be the same.

I think the rate limiter should not need to get passed the indexer_db every time or need to know anything about the storage backends for that matter, just have a different namespace.

File: tests/sentry/sentry_metrics/test_ratelimiters.py [hunk 13]
```diff
@@ -3,6 +3,8 @@
 from sentry.sentry_metrics.indexer.base import KeyCollection
 from sentry.sentry_metrics.indexer.ratelimiters import WritesLimiter
 
+INDEXER_DB = "test-db"
```
Comment (lynnagara) [inline tests/sentry/sentry_metrics/test_ratelimiters.py:6]: I don't think I get this test. There are only 2 valid db's - Postgres and CloudSpanner. What's it trying to test with this fake one?

File: src/sentry/sentry_metrics/configuration.py (hunk 10 above)
Comment (lynnagara) [reply src/sentry/sentry_metrics/configuration.py:60]: Right it's a test now. But if we go ahead with this, won't it become the permanent topic when we cut over? IMO even for a test _CLOUDSPANNER is far more meaningful than dummy since and hardcoded here in the cloudspanner configuration.

File: tests/sentry/sentry_metrics/test_ratelimiters.py (hunk 13 above)
Comment (MeredithAnya) [reply tests/sentry/sentry_metrics/test_ratelimiters.py:6]: oh yeah this is left over, need to remove it

File: src/sentry/sentry_metrics/consumers/indexer/processing.py (hunk 6 above)
Comment (lynnagara) [reply src/sentry/sentry_metrics/consumers/indexer/processing.py]: There are currently two entrypoints to the indexer - one via the product/API and one via ingest. These use cases are quite different but we have to force the interface of the indexer to be the same because of this. Since we are running multiple indexers for ingest I don't think it makes any sense to use the service delegator there as service delegation is mostly useful when you want your users to be able to switch between backends easily. I would propose keeping the service delegator but making it only for Sentry/API use cases and not on the ingest path. In this scenario the in this method the Postgres indexer would just be loaded directly (like the cloudspanner one is) and not via service delegator. What do you think?

File: src/sentry/sentry_metrics/indexer/ratelimiters.py (hunk 12 above)
Comment (MeredithAnya) [reply src/sentry/sentry_metrics/indexer/ratelimiters.py:147]: I think that is what @fpacifici  meant by having the rate limiting partition be a separate attribute in the metrics configuration. I think im gonna open a separate PR to do that for the rate limiter since that can be independent from the backends

File: src/sentry/sentry_metrics/consumers/indexer/processing.py (hunk 11 above)
Comment (lynnagara) [reply src/sentry/sentry_metrics/consumers/indexer/processing.py:67]: You should also change the import to only import the postgres one. Right now it may also be loading the cloudspanner one depending on your settings.

File: src/sentry/sentry_metrics/configuration.py (hunk 10 above)
Comment (fpacifici) [reply src/sentry/sentry_metrics/configuration.py:60]: nit: There does not seem to be a lot of value in assigning `settings.KAFKA_SNUBA_GENERICS_METRICS_DUMMY` to a variable and then using that variable when you instantiate the Config object. It actually makes the code harder to read.

File: src/sentry/sentry_metrics/configuration.py [hunk 14]
```diff
@@ -47,4 +55,30 @@ def get_ingest_config(use_case_key: UseCaseKey) -> MetricsIngestConfiguration:
...
+        _register_ingest_config(
+            MetricsIngestConfiguration(
+                db_backend=CLOUDSPANNER_DB,
+                input_topic=settings.KAFKA_INGEST_METRICS,
+                output_topic=dummy_output_topic,
+                use_case_id=UseCaseKey.RELEASE_HEALTH,
+                internal_metrics_tag="release-health-spanner",
+                writes_limiter_cluster_options=settings.SENTRY_METRICS_INDEXER_WRITES_LIMITER_OPTIONS,
+            )
+        )
```
Comment (fpacifici) [inline src/sentry/sentry_metrics/configuration.py:71]: We do not plan to run a cloudspanner indexer for release health any time soon.  Up to you if you want to have it here or not. No strong opinion.

File: src/sentry/conf/server.py [hunk 15]
```diff
@@ -2444,6 +2449,8 @@ def build_cdc_postgres_init_db_volume(settings):
     KAFKA_INGEST_PERFORMANCE_METRICS: {"cluster": "default"},
     KAFKA_SNUBA_GENERIC_METRICS: {"cluster": "default"},
     KAFKA_INGEST_REPLAYS_RECORDINGS: {"cluster": "default"},
+    # Metrics Testing Topics
+    KAFKA_SNUBA_GENERICS_METRICS_DUMMY: {"cluster": "default"},
```
Comment (fpacifici) [inline src/sentry/conf/server.py:2453]: Remember this will have to be overridden in getsentry in order to avoid hitting the default cluster which is ingest. 
This will be critical before we run the experimental indexer.

File: src/sentry/sentry_metrics/indexer/ratelimiters.py (hunk 12 above)
Comment (fpacifici) [reply src/sentry/sentry_metrics/indexer/ratelimiters.py:147]: I think making this work will be more complex than just passing the indexer_db.
The Configuration object references a structure in server.py which contains the redis connection settings. 
If the test indexer worked on a different cluster that would be enough, But this is unlikely. We will likely share the cluster so we need to implement a way in the rate limiter initialization to provide a prefix in the key as well and to configure that via settings.

**Review by fpacifici — COMMENTED** [review]
Please see the comments in line

File: src/sentry/sentry_metrics/indexer/ratelimiters.py (hunk 12 above)
Comment (fpacifici) [reply src/sentry/sentry_metrics/indexer/ratelimiters.py:147]: More context: 
- now each instance of the indexer creates a number of rate limiters objects (one per use case id). The configuration for each of them in provided by the `Configuration` object. So we can have multiple of them in the same indexer per use case id. Each can reference a different cluster .
- Name spacing in the rate limiter is defined via the `Quota` object per use case id. So the use case id ends up in the key thus guaranteeing the name spacing https://github.com/getsentry/sentry/blob/master/src/sentry/sentry_metrics/indexer/ratelimiters.py#L33-L58
- So we have a mix of rate limiter configuration set at initialization and of parameters passed when the rate limiter is invoked, which is suboptimal. This most likely happened because the `Configuration` abstraction was added at the same time as the rate limiter was being developed.

So there is some refactoring needed here to solve your use case. 
- we cannot rely on the db_id alone and hope that the rate limiter configuration we get from the `Configuration` object will give us a separate name space. That config only contains the cluster to connect to. Which is likely going to be the same.
- we cannot rely on the use case id to be different. It will certainly be the same. 

So there definitively need to be something to add to the `Configuration` object to identify the rate limiter namespace. 
We can temporarily use the db id for this PR but we should have a separate one as I mentioned in my comment last week. 
Then, this string that comes from the config needs to be added to the redis keys of our rate limiter. We will have two options:
- Adding the prefix to the quota or request quota here. You can add a prefix override there. https://github.com/getsentry/sentry/blob/master/src/sentry/sentry_metrics/indexer/ratelimiters.py#L33-L80
- make changes to the basic rate limiter implementation to add a prefix at initialization. I think this refactoring is beyond the scope as all prefix overrides are already defined as part of the Quota or RequestQuota object and not at initialization.

**Review by fpacifici — COMMENTED** [review]
Please see Lyn's comments otherwise it is ok.
I would remove everything related to rate limiter name spacing from here as it would not be complete, so it would be hard to understand. Let's use https://github.com/getsentry/sentry/pull/37828/files for that.

File: src/sentry/sentry_metrics/configuration.py (hunk 10 above)
Comment (lynnagara) [reply src/sentry/sentry_metrics/configuration.py:74]: Actually, I think there is a benefit to defining in settings. It's because the cluster mapping must be defined there. So it keeps those two things together.

Comment (MeredithAnya) [reply src/sentry/sentry_metrics/configuration.py:74]: Yeah Im going to keep in settings, but it doesn't need to be assigned to a variable.

Comment (lynnagara) [reply src/sentry/sentry_metrics/configuration.py]: It needs to be a variable if we want to import it here. Unless you are suggesting just duplicating the value?

**Comment by MeredithAnya** [comment]
@fpacifici @lynnagara @untitaker I had a test that I fixed by added an `IndexerStorage.MOCK` option to basically be able to use the mock indexer in the processes message tests without using the `LazyServiceWrapper`. Are we all on the same page about not using the `LazyServiceWrapper` for the consumer? (and keeping it to be used for the product) Cause that's what this PR is doing.

File: src/sentry/sentry_metrics/consumers/indexer/processing.py [hunk 16]
```diff
@@ -1,18 +1,28 @@
...
 
 logger = logging.getLogger(__name__)
 
+STORAGE_TO_INDEXER = {
+    # TODO(add cloudspanner options)
+    IndexerStorage.CLOUDSPANNER: lambda: CloudSpannerIndexer(**settings.CLOUDSPANNER_OPTIONS),
+    IndexerStorage.POSTGRES: PostgresIndexer(),
+    IndexerStorage.MOCK: MockIndexer(),
+}
+
```
Comment (untitaker) [inline src/sentry/sentry_metrics/consumers/indexer/processing.py:22]:
```suggestion
_MOCK = MockIndexer()
_POSTGRES = PostgresIndexer()

STORAGE_TO_INDEXER = {
    # TODO(add cloudspanner options)
    IndexerStorage.CLOUDSPANNER: lambda: CloudSpannerIndexer(**settings.CLOUDSPANNER_OPTIONS),
    IndexerStorage.POSTGRES: lambda: _POSTGRES,
    IndexerStorage.MOCK: lambda: _MOCK,
}

```

otherwise this will not work, because some of your dict values are callables that return indexers (cloud spanner), while others are actual instances (pg/mock)

Comment (MeredithAnya) [reply src/sentry/sentry_metrics/consumers/indexer/processing.py:22]: oh wait, I don't think cloudspanner needs to be a lamda tho right? the options are static

Comment (untitaker) [reply src/sentry/sentry_metrics/consumers/indexer/processing.py]: yeah but they may not be set, right? if you made it static, having a valid cloudspanner config is now unconditionally required

File: src/sentry/sentry_metrics/consumers/indexer/processing.py (hunk 6 above)
Comment (MeredithAnya) [reply src/sentry/sentry_metrics/consumers/indexer/processing.py]: Updated to have the `process_messages` function choose the indexer via the `STORAGE_TO_INDEXER` mapping instead of the `LazyServiceWrapper`

File: src/sentry/sentry_metrics/consumers/indexer/processing.py (hunk 16 above)
Comment (MeredithAnya) [reply src/sentry/sentry_metrics/consumers/indexer/processing.py]: yeah yeah you are le correct, updated with your suggestion :)

File: src/sentry/sentry_metrics/consumers/indexer/processing.py [hunk 17]
```diff
@@ -1,18 +1,33 @@
...
 logger = logging.getLogger(__name__)
 
+_MOCK = MockIndexer()
+_POSTGRES = PostgresIndexer()
+
+STORAGE_TO_INDEXER: Mapping[IndexerStorage, Callable[[], StringIndexer]] = {
+    # TODO(add cloudspanner options)
+    IndexerStorage.CLOUDSPANNER: lambda: CloudSpannerIndexer(**settings.CLOUDSPANNER_OPTIONS),
+    IndexerStorage.POSTGRES: lambda: _POSTGRES,
+    IndexerStorage.MOCK: lambda: _MOCK,
```
Comment (untitaker) [inline src/sentry/sentry_metrics/consumers/indexer/processing.py:25]: to summarize the discussion in slack, i still think it would be cleaner if there was a single `IndexerStorage.PRODUCT` that uses the globally configured instance, instead of having an enum that has to enumerate all possible backends

File: src/sentry/sentry_metrics/configuration.py [hunk 18]
```diff
@@ -57,4 +73,41 @@ def get_ingest_config(use_case_key: UseCaseKey) -> MetricsIngestConfiguration:
...
+                input_topic=settings.KAFKA_INGEST_PERFORMANCE_METRICS,
+                output_topic=settings.KAFKA_SNUBA_GENERICS_METRICS_CS,
+                use_case_id=UseCaseKey.PERFORMANCE,
+                internal_metrics_tag="perf-spanner",
+                writes_limiter_cluster_options=settings.SENTRY_METRICS_INDEXER_WRITES_LIMITER_OPTIONS_PERFORMANCE,
+                writes_limiter_namespace=PERFORMANCE_CS_NAMESPACE,
+            )
+        )
+
+    if db_backend == IndexerStorage.MOCK:
```
Comment (untitaker) [inline src/sentry/sentry_metrics/configuration.py:100]: can you add a code comment here that we're not cachign the mock indexer in order to provide test isolation?

**Review by untitaker — COMMENTED** [review]
minor nits. i don't think i'm confident enough to give final approval

File: src/sentry/sentry_metrics/consumers/indexer/processing.py [hunk 19]
```diff
@@ -33,12 +48,14 @@ def process_messages(
     The value of the message is what we need to parse and then translate
     using the indexer.
     """
-    batch = IndexerBatch(use_case_id, outer_message)
+    batch = IndexerBatch(config.use_case_id, outer_message)
 
     org_strings = batch.extract_strings()
 
+    indexer = STORAGE_TO_INDEXER[config.db_backend]()
```
Comment (fpacifici) [inline src/sentry/sentry_metrics/consumers/indexer/processing.py:55]: Hold on, you are reinstantiating the indexer object for every message batch you receive. 
That is not needed and actually can be damaging if the indexer instance does something expensive during the instantiation.
Why do you need to make the instantiation of the indexer a Callable ?

File: src/sentry/sentry_metrics/indexer/ratelimiters.py [hunk 20]
```diff
@@ -103,27 +104,31 @@ def __exit__(self, exc_type: Any, exc_value: Any, traceback: Any) -> None:
...
         return self.rate_limiters[namespace]
 
     @metrics.wraps("sentry_metrics.indexer.check_write_limits")
     def check_write_limits(
-        self, use_case_id: UseCaseKey, namespace: str, keys: KeyCollection
+        self,
+        use_case_id: UseCaseKey,
+        namespace: str,
+        db_backend: IndexerStorage,
+        keys: KeyCollection,
```
Comment (fpacifici) [inline src/sentry/sentry_metrics/indexer/ratelimiters.py:131]: I thought the idea of the namespace was to avoid the rate limtiers class to know about the DB backend. It does not seem right that a rate limiter cares whether it is rate limiting spanner or postgres. The rate limiter should work in the same way independently.
What is the reason to introduce the dependency on the DB backend ?

Comment (MeredithAnya) [reply src/sentry/sentry_metrics/indexer/ratelimiters.py:131]: So the problem I ran into was that we get the config in the ratelimiter https://github.com/getsentry/sentry/blob/master/src/sentry/sentry_metrics/indexer/ratelimiters.py#L119 so that we can get the redis options, if the ratelimiter doens't already exist. 

I feel like it could make more sense to initialize the writes_limiter with the namespace and options, or maybe have like an add_namespace method on the `WritesLimiter` so that when we call `check_write_limits` we aren't passing the `db_backend`.

File: src/sentry/sentry_metrics/consumers/indexer/processing.py (hunk 19 above)
Comment (MeredithAnya) [reply src/sentry/sentry_metrics/consumers/indexer/processing.py:55]: I was trying to make this work with `CLOUDSPANNER` initialization, we talk a little about it in https://github.com/getsentry/sentry/pull/37745#discussion_r948329920

Comment (fpacifici) [reply src/sentry/sentry_metrics/consumers/indexer/processing.py:55]: I see the reasoning but it won't be ok. 
If you see how the initialization in Nikhar's PR 
https://github.com/getsentry/sentry/pull/37802/files#diff-6263481e823d6e3fd17a3d6f4a7ffc84b417dd9f1e15db83cc647937536b0399R97-R100
When you initialize the spanner client we establish the TCP connection.
We need to reuse that connection, we cannot re-establish it for every batch..
We generally process tens of batches per second, we cannot establish and destroy the TCP connection each time. Besides the overhead, I would not be surprised if Google throttled us.

So, since the indexer instance is not a static singleton anymore we need an alternative way.
Options:
- Instantiate the indexer in the consumer and pass it to the `process_messages` function.
- Turn the message processing logic into a class so that the indexer can be a local attribute.

File: src/sentry/sentry_metrics/indexer/ratelimiters.py (hunk 20 above)
Comment (fpacifici) [reply src/sentry/sentry_metrics/indexer/ratelimiters.py:131]: I see. 
I think the issue is here 
https://github.com/getsentry/sentry/blob/master/src/sentry/sentry_metrics/indexer/ratelimiters.py#L185

This class is meant to exist in one instance only.
Obviously now we have multiple instances at the same time so having one static instance does not work anymore and we need to be able to return the right rate limiter when we try to write.

> I feel like it could make more sense to initialize the writes_limiter with the namespace and options

I think this is right, a good idea would be to move the options in the initialization of the `WritesLimiter` class and then have a factory that returns the right instance based on the config.

**Review by fpacifici — APPROVED** [review]
